src/DI/ServiceLookup/CallSiteChain.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal class CallSiteChain
+    {
+        private readonly Dictionary<Type,ChainItemInfo> _callSiteChain;
+
+        public CallSiteChain()
+        {
+            _callSiteChain = new Dictionary<Type, ChainItemInfo>();
+        }
+
+        public void CheckCircularDependency(Type serviceType)
+        {
+            if (_callSiteChain.ContainsKey(serviceType))
+            {
+                throw new InvalidOperationException(CreateCircularDependencyExceptionMessage(serviceType));
+            }
+        }
+
+        public void Remove(Type serviceType)
```

Comment:
Nit: It doesn't look like this is called anywhere.

Reply:
It's used in https://github.com/aspnet/DependencyInjection/pull/592/files#diff-a7622bce2aa4a5b7dbb9e510b8fe8f86R87